lib/luaT/luaT.c

```diff
@@ -471,53 +471,146 @@ void luaT_registeratname(lua_State *L, const struct luaL_Reg *methods, const cha
+ * given class `A.b.c`, sets parent_name to `A`)
+ */
+int luaT_outerparentname(const char *tname, char *parent_name)
 {
   char chars[] = {'.', '\0'};
   size_t n;
   n = strcspn(tname, chars);
-  strncpy(module_name, tname, n);
-  module_name[n] = '\0';
+  strncpy(parent_name, tname, n);
+  parent_name[n] = '\0';
   return tname[n] == '.';
 }
 
-/* Lua only functions */
+/* parent_name must be a buffer at least as big as tname.
+ * If class has a parent, returns true; and, sets
+ * parent name to that of innermost parent (e.g.
+ * given class `A.b.c`, sets parent_name to `b`)
+ */
+int luaT_innerparentname(const char *tname, char *parent_name)
+{
+  int sz = strlen(tname);
+  int tail, head;
+  for(tail = sz-1; tail >= 0 ; tail--)
+    if(tname[tail] == '.') break;
+
+  if (tail == 0) return 0;
+
+  for(head = tail; head >= 1; head--)
+    if(tname[head-1] == '.') break;
+
+  strncpy(parent_name, tname+head, tail-head);
+  parent_name[tail-head] = '\0';
+  return 1;
+}
+
+/* Method for pushing a classes immediate parent to the
+ * stack (e.g. given class `A.b.c`, pushes `b` to the stack)
+ */
+void luaT_getinnerparent(lua_State *L, const char *tname)
+{
+  /* Local variables */
+  char term[256];
+  char chars[] = {'.', '\0'};
+  const char *tname_full = tname; // used for error case
+
+  /* Get outermost table from Lua */
+  int n = strcspn(tname, chars);
+  strncpy(term, tname, n);
+  term[n] = '\0';
+  lua_getglobal(L, term);
+  tname  += n + 1;
+
+  /* Traverse hierarchy down to last table*/
+  n = strcspn(tname, chars);
+  while(n < strlen(tname))
+  {
+    if(!lua_istable(L, -1)){
+      strncpy(term, tname_full, tname - tname_full - 1);
+      term[tname - tname_full] = '\0';
+      luaL_error(L, "while creating metatable %s: bad argument #1 (%s is an invalid module name)", tname_full, term);
+    }
+    strncpy(term, tname, n);
+    term[n] = '\0';
+    lua_getfield(L, -1, term);
+    lua_remove(L, -2);
+    tname += n + 1;
+    n = strcspn(tname, chars); // prepare for next
+  }
+}
+
+
 int luaT_lua_newmetatable(lua_State *L)
 {
+  // -------- Local Variables
   const char* tname = luaL_checkstring(L, 1);
-  char module_name[256];
+  char parent_name[256];
   int is_in_module = 0;
-  is_in_module = luaT_classmodulename(tname, module_name);
+  int L_idx = 0;
 
+  // -------- Argument Checking
   lua_settop(L, 5);
   luaL_argcheck(L, lua_isnoneornil(L, 2) || lua_isstring(L, 2), 2, "parent class name or nil expected");
   luaL_argcheck(L, lua_isnoneornil(L, 3) || lua_isfunction(L, 3), 3, "constructor function or nil expected");
   luaL_argcheck(L, lua_isnoneornil(L, 4) || lua_isfunction(L, 4), 4, "destructor function or nil expected");
   luaL_argcheck(L, lua_isnoneornil(L, 5) || lua_isfunction(L, 5), 5, "factory function or nil expected");
 
-  if(is_in_module)
-    lua_getglobal(L, module_name);
+  // -------- Push immediate parent module to stack
+  is_in_module = luaT_outerparentname(tname, parent_name);
```

Comment:
`lua_getinnerparent` doesn't check that the value it leaves on the stack is a table, right? That's why the additional check below is needed. All I'm saying is that the error message would be confusing if that check fails, because it refers to the outer parent whereas it should refer to the inner parent.

Please correct me if I misunderstood something here...